### ISSUE TITLE
SSL behavior differences between OpenSSL >= 1.1 and pre OpenSSL 1.1

### ISSUE THREAD

**Issue by Bigpet** [chriskohlhoff/asio#156]
I'm talking about the changes introduced in https://github.com/chriskohlhoff/asio/commit/af9770ac493b1cc966766c70b32fb5571b566331 . 

I'm aware that the reason that the version specific methods (like tlsv1,tlsv11,tlsv12) were deprecated in OpenSSL 1.1 but they were not removed yet.

I would much prefer that the behavior would not silently change when switching between OpenSSL but rather lead to an exception once they're fully removed like SSLv2.

But I would prefer something like this: https://github.com/Bigpet/asio/commit/8ad6d13926487f17f899816e63ec30029e239f07 . In my opinion this will lead to more consistent behavior.

I did not exhaustively investigate this issue, I just stumbled on this change so I might be missing something here.

**Comment by Bigpet** [comment]
alternatively https://github.com/chriskohlhoff/asio/commit/af9770ac493b1cc966766c70b32fb5571b566331 could be mostly kept mostly intact and in addition to `::SSL_CTX_new(::TLS_method());` a call to `SSL_CTX_set_min_proto_version` could restrict the versions just like before.

```
  case context::tlsv1:
    handle_ = ::SSL_CTX_new(::TLS_method());
    SSL_CTX_set_min_proto_version(handle_,TLS1_VERSION);
    SSL_CTX_set_min_proto_version(handle_,TLS1_VERSION);
    break;
  case context::tlsv11:
    handle_ = ::SSL_CTX_new(::TLS_method());
    SSL_CTX_set_min_proto_version(handle_,TLS1_1_VERSION);
    SSL_CTX_set_min_proto_version(handle_,TLS1_1_VERSION);
    break;
  case context::tlsv12:
    handle_ = ::SSL_CTX_new(::TLS_method());
    SSL_CTX_set_min_proto_version(handle_,TLS1_2_VERSION);
    SSL_CTX_set_min_proto_version(handle_,TLS1_2_VERSION);
    break;
```

this would avoid using deprecated functionality and mostly keep the behavior the same

**Comment by MarcelRaad** [comment]
@Bigpet The second version sounds good. It looks like the first version doesn't work with an OpenSSL lib compiled with --api=1.1.0, which removes the deprecated methods.

**Comment by chriskohlhoff** [comment]
I have implemented the latter approach in d1e0f2127427acf446e454ac39f32b9483e47e26. Please take a look and let me know if it covers your requirements.

**Comment by Bigpet** [comment]
looks fine to me. I really don't require these old versions. I just appreciate the consistency in behavior.
